Week4/LESSON_PLAN.md

```diff
@@ -0,0 +1,275 @@
+];
+```
+
+#### The normalised way
+
+```js
+const accounts = [
+  {
+    id: ObjectId("507f191e810c19729de86032"),
+    email: "rob@thebugcreator.com",
+    name: "Rob",
+  },
+];
+
+const bugs = [
+  {
+    timestamp: new Date(),
+    page: "about",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("507f191e810c19729de86032"),
+  },
+];
+```
+
+### Exercise
+
+Discuss the differences and what the advantages/disadvantages are to each approach. For example:
+
+Embedded allows for faster querying.
+Normalized allows for less data duplication.
+
+### Essence
+
+There are advantages to both approaches and in the wild you will have to decide which to use every time.
+
+## 2. Advanced MongoDB operations
+
+### Explanation
+
+1. The `sort` command allows you to sort data that you get back from your query.
+2. The `limit` command allows you to limit how many items you get back from your query.
+3. The `aggregate` command allows the combination and calculation of data in one or more collections.
+
+### Example
+
+Let's assume a log database with the following information:
+
+```js
+const bugs = [
+  {
+    timestamp: new Date('2000-06-07T11:24:00'),
+    page: "about",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("507f191e810c19729de86032"),
+  }, {
+    timestamp: new Date('2000-06-06T12:23:00'),
+    page: "about",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("507f191e810c19729de86032"),
+  }. {
+    timestamp: new Date('2000-06-08T12:33:00'),
+    page: "contact",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("e810507f191de86032c19729"),
+  }, {
+    timestamp: new Date('2000-06-06T12:34:00'),
+    page: "home",
+    stacktrace: "", // removed for clarity
+    reporterId: ObjectId("e810507f191de86032c19729"),
+  }
+];
+```
+
+#### Sort
+
+If we want to sort the find query on the timestamp to find the latest bugs we can run the following query:
+
+```js
+client.db("logging").collection("bugs").find().sort({ timestamp: -1 });
+```
+
+#### Limit
+
+The above query will give you back all of the bugs which is not great, as such we can use the limit command as follows:
+
+```js
+client
+  .db("logging")
+  .collection("bugs")
+  .find()
+  .sort({ timestamp: -1 })
+  .limit(10);
+```
+
+This will give only the last 10 bugs, which is more manageable. Note that you can put the `sort` and `limit` commands in any order!
+
+#### Aggregate
+
+Let's say we want to have a count of how many bugs appear per page. To do that we can do the following:
+
+```js
+client
+  .db("logging")
+  .collection("bugs")
+  .aggregate([
+    {
+      $group: {
+        _id: "$page",
+        count: { $count: {} },
+      },
+    },
+  ]);
+```
+
+This will give back an object with the page field in the `_id` field and the number of bugs that were logged on that page is in the `count` field!
+
+### Exercise
+
+Think of some other things to sort on or calculate and write the code to do that.
```

Comment:
This exercise is too vague. We could ask a concrete question like:
`Sort all bugs by pages and then by timestamp`